types/dec_coin.go

```diff
@@ -278,6 +278,20 @@ func (coins DecCoins) SafeSub(coinsB DecCoins) (DecCoins, bool) {
 	return diff, diff.IsAnyNegative()
 }
 
+// Trims any denom amount from coin which exceeds that of coinB,
+// such that (coin.Cap(coinB)).IsLTE(coinB).
+func (coins DecCoins) Cap(coinsB DecCoins) DecCoins {
+	res := make([]DecCoin, len(coins))
+	for i, coin := range coins {
+		minCoin := DecCoin{
+			Denom:  coin.Denom,
+			Amount: MinDec(coin.Amount, coinsB.AmountOf(coin.Denom)),
+		}
+		res[i] = minCoin
+	}
+	return removeZeroDecCoins(res)
```

Comment:
Generally, our coins APIs claim the invariant that any coins (int or dec) input should never contain zero coins.

Reply:
coinsB.AmountOf(coin.Denom) can be 0.

Reply:
Ahh, yes. I sit corrected.